Cython/Compiler/Parsing.py

```diff
@@ -3968,6 +3989,440 @@ def p_cpp_class_attribute(s, ctx):
+        if result:
+            s.next()
+            return MatchCaseNodes.MatchValuePatternNode(pos, value=result, is_is_check=True)
+
+    s.error("Failed to match literal")
+
+
+def p_capture_pattern(s):
+    return MatchCaseNodes.MatchAndAssignPatternNode(
+        s.position(),
+        target=p_pattern_capture_target(s)
+    )
+
+
+def p_value_pattern(s):
+    if s.sy != "IDENT":
+        s.error("Expected identifier")
+    pos = s.position()
+    res = p_name(s, s.systring)
+    s.next()
+    if s.sy != '.':
+        s.error("Expected '.'")
+    while s.sy == '.':
+        attr_pos = s.position()
+        s.next()
+        attr = p_ident(s)
+        res = ExprNodes.AttributeNode(attr_pos, obj=res, attribute=attr)
+    if s.sy in ['(', '=']:
+        s.error("Unexpected symbol '%s'" % s.sy)
+    return MatchCaseNodes.MatchValuePatternNode(pos, value=res)
+
+
+def p_group_pattern(s):
+    s.expect("(")
+    pattern = p_pattern(s)
+    s.expect(")")
+    return pattern
+
+
+def p_sequence_pattern(s):
+    opener = s.sy
+    pos = s.position()
+    if opener in ['[', '(']:
+        closer = ']' if opener == '[' else ')'
+        s.next()
+        # maybe_sequence_pattern and open_sequence_pattern
+        patterns = []
+        if s.sy == closer:
+            s.next()
+        else:
+            while True:
+                patterns.append(p_maybe_star_pattern(s))
+                if s.sy == ",":
+                    s.next()
+                    if s.sy == closer:
+                        break
+                else:
+                    if opener == ')' and len(patterns)==1:
+                        s.error("tuple-like pattern of length 1 must finish with ','")
+                    break
+            s.expect(closer)
+        return MatchCaseNodes.MatchSequencePatternNode(pos, patterns=patterns)
+    else:
+        s.error("Expected '[' or '('")
+
+
+def p_mapping_pattern(s):
+    pos = s.position()
+    s.expect('{')
+    if s.sy == '}':
+        # trivial empty mapping
+        s.next()
+        return MatchCaseNodes.MatchMappingPatternNode(pos)
+
+    double_star_capture_target = None
+    items_patterns = []
+    star_star_arg_pos = None
+    while True:
+        if double_star_capture_target and not star_star_arg_pos:
+            star_star_arg_pos = s.position()
+        if s.sy == '**':
+            s.next()
+            double_star_capture_target = p_pattern_capture_target(s)
+        else:
+            # key=(literal_expr | attr)
+            with tentatively_scan(s) as errors:
+                pattern = p_literal_pattern(s)
+                key = pattern.value
+            if errors:
+                pattern = p_value_pattern(s)
+                key = pattern.value
+            s.expect(':')
+            value = p_pattern(s)
+            items_patterns.append((key, value))
+        if s.sy==',':
+            s.next()
+        else:
+            break
+        if s.sy=='}':
+            break
```

Comment:
I think this spelling is simpler.
```suggestion
        if s.sy != ',':
            break
        s.next()
        if s.sy == '}':
            break  # Allow trailing comma.
```